### PR TITLE
Country plants balance classifier

### PR DESCRIPTION
**Pull request by mortonjt**
Redid all of the country / plants analyses using the new balance classifier.

This contains PLS scatter plots, boxplots, histograms, ROC curves, mean proportion plots and an explanation of how the algorithm works.  If there are any issues with this code - please let me know ASAP.

### CODE REVIEW

File: ipynb/western-nonwestern-balances.ipynb [hunk 1]
```diff
@@ -0,0 +1,891 @@
...
+    "\n",
+    "metadata = pd.read_table('mapping_meta_urban_rural.txt', index_col=0)\n",
+    "metadata = metadata.drop_duplicates()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We'll apply some very relaxed filters.  If an OTU only as 10 reads in the whole study, or only appears in 10 samples, it likely will not be informative.  So we'll filter out these samples. "
```
Comment (wasade) [inline ipynb/western-nonwestern-balances.ipynb]: I dont think the values described here reflect the thresholds used in the code cell

File: ipynb/western-nonwestern-balances.ipynb [hunk 2]
```diff
@@ -0,0 +1,891 @@
...
+    "\n",
+    "ctable = pd.DataFrame(clr(centralize(table+1)), index=table.index, columns=table.columns)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "To just visualize the results, we'll perform Partial Least Squares with two components.\n",
+    "This will verify if there are clear patterns with respect to the types of plants and country analysis."
```
Comment (wasade) [inline ipynb/western-nonwestern-balances.ipynb:130]: this is western/nonwestern though, right?

File: ipynb/western-nonwestern-balances.ipynb [hunk 3]
```diff
@@ -0,0 +1,891 @@
...
...L5C3eDsSkNs9lK82HR4qt1Fy/Z2p63cX\ny+Uyxc9+mVd++j1aq+coGmAtXaKx2mV4coaNpTkmigYU99KwQnRdYaKg02w2tiUJF66uUCjm+dT9\nH98+/9Y57n7oUb7zX/4nxI03oLPMREZjTqhycTmNyUHi2MXQbIbyZbLKBKVMncc+5vH8+gqP/tf/\nA72fryGKIsOTM6wvztKrzaLGNm7X48cvneX4Z2e4WutyQu6gFfs7hSXA8zucqfW4/zbr6s3WaUpH\nkZKEM9/9z+9bkNet3uu3M3T3bks2BvzLcNsC+Vvf+hZ/9Ed/xPz8PPfeey9f//rX78in+MPA1iLn\ne1fZXJ6jt/QGWkqkFSVs2hmmfA9VVZh74XsE1YtklYgrkcILL/+Cu7UVSqaMZ4Fn+XiOhRckFDMa\nmujQ7viYUgjWMpuWQzN2ycgRYxmBnhvRc2IUMeHUhMzfPPkT0maaBysRLTukaMoYqkhCgudaCHHI\n8qKL5MYs/tM30TJFVsszPPD5X6Vr2RiX/p5CKc16dQOcDmEk0QtFUmGAJleIohAhCgm6NbSMgdhr\nEIQRatQjZSiIjf7Xkijw8nM/RepskJ0cp1UXifUiJ/fv54W5FtPDWUzR5/VLV9lfUhnfdwhBEMiF\nEd0rF2msDjMydWB7yhtAl0Vs179pfDZAo9G442L1reyb7mQQ6JdRBzlgALzztfiv//qv+Yu/+Itd\n3/vTP/1TvvSlL/Hiiy9uf6/X65FOp7e/Nk2TpaWltzz+uxna9GHiZsVelJ9maijYVTTDzV13xqNl\nlD0Fmh2L8ZxEOmqxtnAJwWltz3wUsmmWGw7j+RghdIjjGEjIGgpxbnjXebbO8YunQx46UKReuYez\nz1Vpdy0+MebgdQwksU4YRkwUNRRVIyZhvWYQ7w8pZTIUi0Wq15pGfY9jGKqUiKI8LTng5IHDaL1Z\nNKeKVyriXnM28pP+cONUOnODdO76Nfn6dfr153/ISLCAUtQIQhnJdhE7l3j6H7t88rHfeF8LzZsN\n3VUvu/xd9wm+8q8eveFa3k3JxoB/GW5bIP/+7/8+x44d42tf+xrf//73+cY3vsE3vvGN9+va/sXY\n+g969OTD/MOf/y+k6qukwoDltkjT8pkqJtRe/zGLzZAxE0rqFIIo0V14g/3CIrIoocki1doGJS
```
Comment (wasade) [inline ipynb/western-nonwestern-balances.ipynb]: Not western/nonwestern again

Comment (wasade) [inline ipynb/western-nonwestern-balances.ipynb]: note western/nonwestern

Comment (wasade) [inline ipynb/western-nonwestern-balances.ipynb]: "but to that, we" -> "but to do that, we'll"

Comment (wasade) [inline ipynb/western-nonwestern-balances.ipynb]: merge -> melt

Comment (wasade) [inline ipynb/western-nonwestern-balances.ipynb]: note western/nonwestern

File: ipynb/core-balance-classifier.ipynb [hunk 4]
```diff
@@ -0,0 +1,1078 @@
...
+    "taxa = pd.DataFrame(taxa, index=obs_ids, \n",
+    "                    columns=['kingdom', 'phylum', 'class', 'order', \n",
+    "                             'family', 'genus', 'species'])\n",
+    "\n",
+    "# we will propagate the last annotation\n",
+    "def taxa_fix(x):\n",
+    "    suffixes = ['k__', 'p__', 'c__', 'o__', 'f__', 'g__', 's__']\n",
+    "    for i, xi in enumerate(x):\n",
+    "        if xi in suffixes:\n",
+    "            x[i] = x[i-1]\n",
```
Comment (wasade) [inline ipynb/core-balance-classifier.ipynb]: this, and in the western/nonwestern notebook, leads to things this in the summary output:

k__Bacteria,p__Firmicutes,c__Clostridia,o__Clostridiales,f__Christensenellaceae,f__Christensenellaceae,f__Christensenellaceae

...would it be possible for the summarized output to reflect the unclassified ranks and lineage structure? eg:

k__Bacteria,p__Firmicutes,c__Clostridia,o__Clostridiales,f__Christensenellaceae,g__,s__

**Review by wasade — COMMENTED** [review]
i thought i had submitted these earlier but i didn't

File: ipynb/core-balance-classifier.ipynb [hunk 5]
```diff
@@ -0,0 +1,979 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Here we will create balance classifiers to identify differences due to plant diet and country.\n",
+    "\n",
+    "Make sure that the appropriate files are downloaded.  They can be obtained from the ftp link using the following commands\n",
```
Comment (tkosciolek) [inline ipynb/core-balance-classifier.ipynb:9]: specify which environment to use for this notebook

**Review by serenejiang — APPROVED** [review]
This looks good to me!

**Review by tkosciolek — COMMENTED** [review]
looks good overall!

File: ipynb/core-balance-classifier.ipynb (hunk 4 above)
Comment (mortonjt) [reply ipynb/core-balance-classifier.ipynb]: We did this for the plots - in order to have the plots show the last annotated level, we needed to have something like this enabled.

The original taxa summaries are still there, so I don't see a reason why this needs to change.

File: ipynb/western-nonwestern-balances.ipynb (hunk 1 above)
Comment (mortonjt) [reply ipynb/western-nonwestern-balances.ipynb]: Fixed.  Thanks!

**Comment by mortonjt** [comment]
All comments have been addressed!  Thanks!

File: ipynb/core-balance-classifier.ipynb (hunk 4 above)
Comment (wasade) [reply ipynb/core-balance-classifier.ipynb]: This impacts the generated tables of differential sOTUs which have incorrect taxon strings, and which need to be corrected for the submission

File: ipynb/core-balance-classifier.ipynb (hunk 5 above)
Comment (mortonjt) [reply ipynb/core-balance-classifier.ipynb:9]: Done.  Thanks!

File: ipynb/core-balance-classifier.ipynb (hunk 4 above)
Comment (mortonjt) [reply ipynb/core-balance-classifier.ipynb]: Ahhh.  That makes sense.  I fixed the taxonomies so that they are correctly outputted for the final results.

**Pull request merged**